bridge-fungible-token/src/utils.sw

```diff
@@ -37,16 +93,30 @@ pub fn adjust_withdrawal_decimals(val: u64) -> b256 {
 
 /// Make any necessary adjustments to decimals(precision) on the deposited value, and return either a converted u64 or an error if the conversion can't be achieved without overflow or loss of precision.
 pub fn adjust_deposit_decimals(msg_val: b256) -> Result<u64, BridgeFungibleTokenError> {
-    let decomposed = decompose(msg_val);
-    let value = U256::from((decomposed.0, decomposed.1, decomposed.2, decomposed.3));
+    let value = U256::from(decompose(msg_val));
 
     if LAYER_1_DECIMALS > DECIMALS {
-        let adjustment_factor = U256::from((0, 0, 0, 10.pow(LAYER_1_DECIMALS - DECIMALS)));
-        if value.divide(adjustment_factor).multiply(adjustment_factor) == value
-            && (value.gt(adjustment_factor)
-            || value.eq(adjustment_factor))
+        let decimal_diff = LAYER_1_DECIMALS - DECIMALS;
+        //10.pow(19) fits in a u64, but 10.pow(20) would overflow when
+        // calculating adjustment_factor below.
+        // There's no need to check this in adjust_withdrawal_decimals();
+        // if an overflow is going to occur when calculating adjustment_factor,
+        // it will be caught here first.
+        if decimal_diff > 19u8 {
```

Comment:
So, I imagined we'd create a new function that wraps the newly made `bn_mul` function. This function would multiply by any power of 10 and return an error if it overflows.
```
fn shift_decimals_left(bn: U256, d: u8) -> b256 {
  let result: b256 = bn; // clone the big number
  
  // the zero case
  if(d == 0) return result

  // the too large case
  // (there are only 78 decimal digits in a 256bit number)
  if(d > 77) return an error

  // math time
  while(d > 0) {
    if(d<20) {
      //multiply result by 10.pow(d)
      if(overflow) return error
      d = 0;
    } else {
      //multiply result by 10.pow(19)
      if(overflow) return error
      d = d - 19;
    }
  result
}
```

This is also partly why I originally wrote the functions to modify the big number in place (since I figured it would be wrapped by the actual function we use)

Reply:
Once we have a working `shift_decimals_left` and `shift_decimals_right` all we have to do is call one or the other, check that there wasn't an error from it, then check that the result isn't too large for a u64